### PR TITLE
Refactor SelectingItemsControl selection.

### LINKED ISSUE #2004: ListBox does not select properly with repeated items
Only one of the top three lines will be selected wherever you click
```
<ListBox >
    <ListBox.Items xmlns:sys="clr-namespace:System;assembly=mscorlib">
      <sys:String>Hello</sys:String>
      <sys:String>Cruel</sys:String>
      <sys:String>World</sys:String>
      <sys:String>Hello</sys:String>
      <sys:String>Cruel</sys:String>
      <sys:String>World</sys:String>
      <sys:String>Hello</sys:String>
      <sys:String>Cruel</sys:String>
      <sys:String>World</sys:String>
      <sys:String>Hello</sys:String>
      <sys:String>Cruel</sys:String>
      <sys:String>World</sys:String>
      <sys:String>Hello</sys:String>
      <sys:String>Cruel</sys:String>
      <sys:String>World</sys:String>
      <sys:String>Hello</sys:String>
      <sys:String>Cruel</sys:String>
      <sys:String>World</sys:String>
      <sys:String>Hello</sys:String>
      <sys:String>Cruel</sys:String>
      <sys:String>World</sys:String>
    </ListBox.Items>

   
    <ListBox.ItemsPanel>
      <ItemsPanelTemplate>
        <StackPanel Spacing="2" />
      </ItemsPanelTemplate>
    </ListBox.ItemsPanel>
    <ListBox.ItemTemplate >
      <DataTemplate >
        <StackPanel Orientation="Horizontal">
          <StackPanel Orientation="Horizontal" Height="24">
            <TextBlock Width="80" Text="{Binding}"></TextBlock>
          </StackPanel>
        </StackPanel>
      </DataTemplate>
    </ListBox.ItemTemplate>
  </ListBox>
```

### LINKED ISSUE #2574: Property SelectedIndex does not change when an item was added to the items before the selected item
The SelectedIndex property doesn't changed when insert an item before the SelectedIndex into the INotifyCollectionChanged binding to the items property of a SelectingItemsControl .

There should be a check in the add part of the ItemsCollectionChanged function in the SelectingItemsControl, which could be like
```c#
                case NotifyCollectionChangedAction.Add:
                    if (AlwaysSelected && SelectedIndex == -1)
                    {
                        SelectedIndex = 0;
                    } 
                    else if (e.NewStartingIndex <= SelectedIndex)
                    {
                        SelectedIndex += e.NewItems.Count;
                    }

                    break;
```
or simply set the SelectedIndex to the index of SelectedItem like the move and reset action.

https://github.com/AvaloniaUI/Avalonia/blob/6fdec1649c5d88a6d4a407ae78c3a3e8d57a8ff3/src/Avalonia.Controls/Primitives/SelectingItemsControl.cs#L351-L357

### LINKED ISSUE #2660: Selection is shrunk to a single item under the cursor on right click on ListBox
But it should be several items. I've encountered this bug in AvaloniaILSpy: https://github.com/icsharpcode/AvaloniaILSpy/issues/13

See the implementation of multiselection here: https://github.com/AvaloniaUI/Avalonia/issues/206

### PR DESCRIPTION
**Pull request by grokys**
## What does the pull request do?

As described in #2004, `SelectingItemsControl` did not correctly handle duplicate items. This was because `SelectedItems` was used as the canonical source of selection information, which meant that it was impossible to determine _which_ of the duplicate items was selected.

Instead of using `SelectedItems` as the canonical source of selection information, use an internal `Selection` class which keeps track of the selection in terms of item indexes.

## Note

This PR removes the need for the `SetAndRaise` overload that takes a callback. I think the same technique can be used in `TreeView` and when that's done, we can remove that `SetAndRaise` overload, making #2362 a lot simpler (cc: @MarchingCube).

## Checklist

- [x] Added unit tests (if possible)?
- [ ] Added XML documentation to any related classes?
- [ ] Consider submitting a PR to https://github.com/AvaloniaUI/Avaloniaui.net with user documentation

## Fixed issues

Fixes #2004 
Fixes #2574 
Fixes #2660

### CODE REVIEW

**Comment by grokys** [comment]
cc: @ahopper @Kermalis @FoggyFinder @ImNotYzq - sorry, this took a bit longer than expected! Could you check if this fixes your issues?

**Comment by ahopper** [comment]
this works for me, although I had to tweak the repro to make it work with XamlIl.

**Comment by grokys** [comment]
There's actually still a few things wrong here related to updating the selected indexes when inserting and removing items. Working on fixing now.

**Comment by grokys** [comment]
Ok, selection should hopefully now correctly react to all changes in the underlying items.

**Comment by ImNotYzq** [comment]
this works for me

File: src/Avalonia.Controls/Primitives/SelectingItemsControl.cs [hunk 1]
```diff
@@ -103,6 +104,7 @@ public class SelectingItemsControl : ItemsControl
 
         private static readonly IList Empty = Array.Empty<object>();
 
+        private Selection _selection = new Selection();
```
Comment (MarchingCube) [inline src/Avalonia.Controls/Primitives/SelectingItemsControl.cs]: Nit: `readonly` modifier? As this is never written to again.

File: src/Avalonia.Controls/Primitives/SelectingItemsControl.cs [hunk 2]
```diff
@@ -981,5 +1068,104 @@ private void UpdateFinished()
                 SelectedItems = _updateSelectedItems;
             }
         }
+
+        private class Selection : IEnumerable<int>
+        {
+            private List<int> _list = new List<int>();
```
Comment (MarchingCube) [inline src/Avalonia.Controls/Primitives/SelectingItemsControl.cs]: Nit: Both could be `readonly`.

Comment (grokys) [reply src/Avalonia.Controls/Primitives/SelectingItemsControl.cs]: Fixed the `readonly`s. `_set` can't be readonly though because it's reassigned.

**Comment by grokys** [comment]
Also added a fix for #2660 to this PR.

**Comment by Kermalis** [comment]
I saw "Fix clearing selection when clicking on selected item" and want to ask, will that test and fix still respect SelectionMode.AlwaysSelected?

**Comment by grokys** [comment]
@Kermalis do you mean that `SelectedMode.AlwaysSelected` is currently broken? Is there an issue?

**Comment by Kermalis** [comment]
No, I don't mean that, I was just asking if it would respect it with the new changes, just in case it was overlooked
Sorry for the confusion

**Comment by grokys** [comment]
It should respect `AlwaysSelected`, yes. If not, it's a bug.

**Pull request merged**